File: Arrays/Rotate_2D_Matrix.cpp
// Rotate an n x n matrix 90 degrees clockwise

#include <iostream>
#include <iomanip>
using namespace std;

int main() {
	int n;
	cin >> n;
	int m[n][n];
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> m[i][j];
		}
	}

	// First take transpose
	for (int i = 0; i < n; i++) {
		for (int j = i + 1; j < n; j++) {
			swap(m[i][j], m[j][i]);
		}
	}

	// Now reverse order of columns
	for (int i = 0, j = n - 1; i < j; i++, j--) {
		for (int k = 0; k < n; k++) {
			swap(m[k][i], m[k][j]);
		}
	}
	// If 90 degrees anticlockwise required, reverse rows after transpose

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cout << setw(3) << m[i][j] << ' ';
		}
		cout << '\n';
	}
}

File: Java/src/Arrays/Rotate2DMatrix.java
package Arrays;

import FastScanner.FastScanner;

import java.io.BufferedOutputStream;
import java.io.PrintWriter;

/*
    Rotate a NxN matrix by 90 deg anticlockwise
    Do this in-place
 */
public class Rotate2DMatrix {
    static PrintWriter out;

    public static void rotate(int[][] a) {
        int n = a.length;
        // Take transpose
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                int t = a[i][j];
                a[i][j] = a[j][i];
                a[j][i] = t;
            }
        }
        
        // Reverse all columns
        // If we needed to rotate by 90 deg clockwise, we would reverse all rows
        for (int j = 0; j < n; j++) {
            for(int i=0, k=n-1; i<k; i++, k--) {
                int t = a[i][j];
                a[i][j] = a[k][j];
                a[k][j] = t;
            }
        }
    }

    public static void main(String[] args) {
        FastScanner fs = new FastScanner();
        out = new PrintWriter(new BufferedOutputStream(System.out));

        int n = fs.nextInt();
        int[][] a = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = fs.nextInt();
            }
        }
        
        rotate(a);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                out.print(a[i][j] + " ");
            }
            out.println();
        }
        out.close();
    }
}
